### ISSUE TITLE
bug: maximize mode - cursor position not consistent

### ISSUE THREAD

**Issue by vdegenne** [LazyVim/LazyVim#4934]
### Did you check docs and existing issues?

- [x] I have read all the LazyVim docs
- [x] I have updated the plugin to the latest version before submitting this issue
- [x] I have searched the existing issues of LazyVim
- [x] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.10.2

### Operating system/version

Fedora 41

### Describe the bug

when using maximize window mode, switching from one window to another will cause the cursor to relocate in wrong locations.

https://github.com/user-attachments/assets/b4c95819-ba94-4944-b477-565a2c2228e7

### Steps To Reproduce

1. open two window, 2 different buffers
2. activate maximize window (`<leader>wm`)
3. alternate window using `c-j` and `c-k`

The cursor is changing places.

### Expected Behavior

The cursor should always stay on the line it was one.

### Repro

```lua

```

**Comment by folke** [comment]
I'm not sure I understand what you mean.
The `maximize` functionality just does this under the hood:

```lua
        set("winwidth", 999)
        set("winheight", 999)
        set("winminwidth", 10)
        set("winminheight", 4)
```

So your current window will become the large window.

Everything in your video looks like it should

**Comment by vdegenne** [comment]
I am talking about the cursor position, if you look in the video, when I switch from one maximized window to another and go back again in the previous one the cursor/caret is no longer on the same line, which is really a bug.

**Comment by folke** [comment]
LazyVim does not change the cursor position. Like I said, we just set those options.

If your original window becomes smaller, it can happen that the cursor position changes.
Things like an animate plugin or `vim.o.splitkeep = screen`, might affect this as well.

Either way, this is not a LazyVim bug...

**Comment by vdegenne** [comment]
ok `vim.opts.splitkeep = "cursor"` fixed it

**Comment by vdegenne** [comment]
@folke When I close a session with Maximize feature left on and reopen nvim and press `s` to retrieve the last state, the windows are still "maximized" but the Maximize feature is turned off and using `<leader>w=` to even the window does nothing. Is there something wrong in my config or do you have the same problem? I can open a new issue for that.

**Comment by folke** [comment]
fixed on main

**Comment by dpetka2001** [comment]
Is this because of the change of `get/set` into methods in `Snacks.toggle` as opposed to just being fields of the toggle before?

**Comment by folke** [comment]
yep, indeed. All good now

**Comment by vdegenne** [comment]
@folke Is there a reason you enforce 4 for the minheight? That'd be great to have a way to be able to customize these values.
Afaik this will overwrite vim options:

https://github.com/LazyVim/LazyVim/blob/9ec253b9b6bf9673f610035d741a345d6be7a44f/lua/lazyvim/util/ui.lua#L47-L53

**Comment by dpetka2001** [comment]
Yeah, in my opinion the `min` values should just not be set at all and this way they would end up whatever the user chooses them to be or the default values. I end up utilizing `<C-w>_` for maximum height instead of `<leader>wm` for this reason. Personally, I prefer to have smaller `min` values because of the smaller screen real estate in my laptop.

**Comment by vdegenne** [comment]
@dpetka2001 Agreed, for min values. Personally I use Maximize + winbar, so the windows act like tab/drawers but stacked horizontally (on a vertical screen). Combined with reuse_win in most actions I can easily remember how stacked windows are and can easily travel using `<c-j>` and `<c-k>`.

**Comment by folke** [comment]
I created a new zen plugin for snacks that also includes zoom functionality.
You can check out the PR here: https://github.com/folke/snacks.nvim/pull/196

When you enable `zoom`, your editor will look the same as if all splits would have been removed (but of course they're still there).

Let me know what you think!

**Comment by dpetka2001** [comment]
From minimal testing it, it looks great. Do you also have plans to incorporate something like `twilight.nvim` maybe? Or is it out of scope? 

To clarify, this is not a request, just a pure question. You do what you think best. I'll keep using `twilight.nvim` if out of scope. No prob.

**Comment by folke** [comment]
I was indeed :)
Going to do that next. Will be simpler than twilight.nvim.

I'm thinking of just using nvim-treesitter textobjects for this, or hl captures.

**Comment by folke** [comment]
What do you think of the zoom functionality?
My only issue with it right now, is that it's indistinguishable from a real window, so users may get confused and forget they're in a zoom window? Or do you think that's ok? An option could be to add some sort of indicator that the window is zoomed.

**Comment by dpetka2001** [comment]
> My only issue with it right now, is that it's indistinguishable from a real window, so users may get confused and forget they're in a zoom window?

Maybe some kind of indicator in the statusline? Or something else if you think of something better would indeed be valuable info.

**Comment by vdegenne** [comment]
> I created a new small zen plugin for snacks that also includes zoom functionality. You can check out the PR here: [folke/snacks.nvim#196](https://github.com/folke/snacks.nvim/pull/196)
> 
> When you enable `zoom`, your editor will look the same as if all splits would have been removed (but of course they're still there).
> 
> Let me know what you think!

Unless I am not using it properly zen is just zooming the current window but for instance if `<c-j>` is pressed it will not change the window and retain the zoom. But I still like the idea of this feature, I'll def be using it if it's bind to something like `<leader>wz` by default!

**Comment by vdegenne** [comment]
Also maybe that's a silly suggestion but why not turning twilight on by default when zen mode is entered, so the user knows for a fact they are zen, you can always give them an option to disable that default behavior, in that case they are responsible to remember if they are in zen mode or not.

**Comment by dpetka2001** [comment]
If i remember correctly pressing `<c-j>` took me out of zoom mode. Not at home right now, so will check again to make sure when i get back in a couple of hours.

**Comment by vdegenne** [comment]
@folke Ah my bad, I am using a vertical screen so zoom and zen did look the same, but now I know what you mean. Using a little fullscreen icon in the corner is a great suggestion from @dpetka2001

**Comment by folke** [comment]
I went a bit overboard with a couple of new snacks plugins :)

Check it out at https://github.com/folke/snacks.nvim/pull/196
 Let me know what you think!

**Comment by dpetka2001** [comment]
Oh, was a bit baffled with `animate`. Thought it was something like `mini.animate`, but then I read more carefully that it was `(library)` and was used by `dim` and `indent` 🤦 

I'm using that branch for now. On an unrelated note, what does this number mean in the `Snacks.profiler`?
![Image](https://github.com/user-attachments/assets/38354f58-47d3-403b-8386-47dfc14a582b)

How many times the function has been called?

**Comment by dpetka2001** [comment]
Also have you given any thought about some special info when the window is zoomed maybe? Some kind of extmark or maybe some kind of info that can be called for example in Lualine and show a letter `M` for example? 

Also a little problem I found when toggling zoom on/off. Screencast following
https://github.com/user-attachments/assets/0cca85d4-a4b7-401a-b3bb-3ed03940d275

It only happened to me so far when I was zooming the bottom window some times. If I zoom the top window then I haven't observed any weirdness so far.

**Comment by dpetka2001** [comment]
I would also like to note that I noticed this side-effect only when using `<leader>Z` for toggling zoom on/off. If I use for example `<c-j>` or `<c-h>` to exit zoom mode and then close the right/bottom window, I haven't observed this happening.

**Comment by folke** [comment]
hmm, I don't seem to be able to reproduce that problem. Does it get fixed when you do `:redraw!` afterwards?

**Comment by folke** [comment]
The number is indeed the amount of times it was called.
The time numbers for methods like `on_win` can be a bit off though, since those methods are fast and very frequent.
The overhead of the profiler is going to be substantial for those cases.

For indent, there's a hidden `opts.debug` that you can enable and when you do, you'll see that pretty much everything in `on_win` is cached at all times, so it should be very fast.

Since I work with window-local extmarks, you can have the same file open in a split for example and see the correct scopes for each window.

Same with side scrolling. Side scrolling the same buffer in one of the two windows, messes up the indent guides with indentblankline.

**Comment by folke** [comment]
How does this look for zoom indicator?

![Image](https://github.com/user-attachments/assets/dae5799a-441a-4963-8b46-5489746abf15)

**Comment by dpetka2001** [comment]
I believe it the visual artifact I'm seeing is due to `mini.animate`. I disabled it and can't reproduce it any more no matter how many times I tried. Previously with `mini.animate` it happened 1 out of 3 times maybe (sometimes even 1 out of 2). I remember seeing an issue with `mini.animate` with the same thing about Neotree windows and it was due to a bug in Neovim with sidebar windows? I don't remember exactly to be honest. 

Bottom line is I disabled `mini.animate` and the visual artifact is not there any more. And no `:redraw!` wasn't working either. I had either to recreate the split window and zoom in and out again and then close the window or restart Neovim.

**Comment by dpetka2001** [comment]
Screencast with `mini.animate` enabled/disabled

https://github.com/user-attachments/assets/3e82c08b-1772-45c6-b30b-0cb40cc6d52f

**Comment by dpetka2001** [comment]
I ended up with `opts.close.enable = false` for `mini.animate`, which solves the problem without completely disabling the plugin.

**Comment by dpetka2001** [comment]
For the `indent` plugin it would be good to have a feature to disable the current scope indent just like we do for `indent-blankline` when we enable `mini.indentscope`. Currently they overlap and don't show correct indents. Screencast following

https://github.com/user-attachments/assets/47fa0818-b1c4-4f11-946c-97dc618dc2e1

**Comment by folke** [comment]
You can get pretty much the exact same indents with snacks by just disabling treesitter scopes.

`opts.indent.scope.treesitter.enabled = false`.

Apart from that, there shold indeed also be an option to disable scopes.

**Comment by dpetka2001** [comment]
I've tried the option you suggested, but the behavior stays the same from what I can observe.

https://github.com/user-attachments/assets/2420ce43-d933-4d48-8cfd-7bc83b265948

It seems that for the first child the scope indent is not correct. Notice the difference when I disable `snacks_indent`, how `mini.indentscope` shows the current scope indent and how it is shown when `snacks_indent` is enabled.

**Comment by dpetka2001** [comment]
Notice in 17sec of the screencast, when I go with my cursor to `styles.notification` how it shows the current scope indent and how it shows it when I go to `styles["notification.history"]`. In my eyes (and I might be wrong about this) it seems to not calculate the current scope indent for the first child correctly. Whereas from the second child onwards it shows it correctly.

**Comment by dpetka2001** [comment]
Also, I think that currently the setup for `indent` doesn't merge the user settings?? Or am I wrong?

**Comment by folke** [comment]
I was getting the wrong options, so yes that didnt do anything :)

Please check the updated changes.

As for the original issue, that's something I still need to fix for treesitter. The `width_edge` impl is not entirely correct.

But with treesitter disabled you should now get pretty much the same results as with indentscope.

**Comment by dpetka2001** [comment]
Yes, now it's better. Most indents look like `mini.indentscope` with some exceptions. What I mentioned earlier for `styles.notification` and `styles["notification.history"]` as also can be seen in mark 17sec of the last screencast I posted, still remains the same. 

You know better for any possible solution. A setting to disable the current scope indents (like `indent-blankline` has) would also be welcome.

**Comment by folke** [comment]
Set `opts.indent.scope.min_size = 1`

**Comment by folke** [comment]
Is 2 by default, which is good. But `edge` detection doesn't yet take that into account.
Will fix, but for now, just set that to `1`

**Comment by folke** [comment]
Fixed now. So you can leave it at 2

**Comment by dpetka2001** [comment]
Thank you for your suggestion. Now it looks like how I am normally used to seeing the indents 😄

**Comment by dpetka2001** [comment]
Ok, I just pulled again and removed `min_size` and with treesitter disabled it looks the same now.

**Comment by dpetka2001** [comment]
Another thing of note is that when you use zen/zoom mode with the suggested mappings in the PR, when you toggle them off it breaks the mapping `Ctrl-^` for the alternate buffer. If instead of `<leader-z>/<leader>Z` you use `<c-h>` or any other mapping to move into a window or if you open any other window (Neotree, Fzflua picker) then you are able to use it.

**Comment by folke** [comment]
ok, will check!

I also just added a `scroll` plugin for smooth scrolling similar to `mini.animate`.

**Comment by dpetka2001** [comment]
Seems like in `WinClosed` event a `win:close()` is needed?? That solved it for me locally although I'm not sure if there might be any other side effects.

**Comment by dpetka2001** [comment]
I'm getting error with latest pull
```lua
Error detected while processing UIEnter Autocommands for "*":
Error executing lua callback: ...e/jrn23/projects/plugins/snacks.nvim/lua/snacks/init.lua:123: attempt to index a nil value
stack traceback:
	...e/jrn23/projects/plugins/snacks.nvim/lua/snacks/init.lua:123: in function <...e/jrn23/projects/plugins/snacks.nvim/lua/snacks/init.lua:121>
```
If I put `scroll` in my snacks config it goes away.

**Comment by dpetka2001** [comment]
Tnx!! 😄

**Comment by folke** [comment]
How useful would an icons plugin be? (not like mini.icons or devicons)

You can get icons like: (autocompletion works of course)
```lua
local icon = Snacks.icons.fa.sun
```

Quickly find/yank a nerd icons:

![Image](https://github.com/user-attachments/assets/93c7f1e9-400d-48ca-8ff1-ed6caae57c43)

**Comment by dpetka2001** [comment]
So, something more like the `cmp-emoji` source, but builtin to snacks? I don't use that much icons myself to be honest, but I don't think it'll hurt. Maybe ask more people about that in an issue at snacks repo to gather if there's interest? @vdegenne Would you like to share your opinion about the icons plugin?

**Comment by vdegenne** [comment]
I don't use icons much too, but If I ever wanted to find an icon without leaving my editor then the second screenshot @folke shared in his last response would be the way, telescope selection (ideally lazy loaded) and direct pasting of the actual icon where the cursor is in the buffer.
Having a lua icons table is interesting too, my only concern is performance, not sure how autocompletion tools are going to index/handle ten thousands of items, especially on old cpu or slow remote ssh connections.
